### PR TITLE
Refactor & Typescript

### PR DESCRIPTION
**Pull request by mahboubii**

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -1,20 +1,13 @@
...
-    "stream-analytics"
-  ],
-  "license": "Apache-2.0",
-  "ignore": [
-    "tests",
-    "vendor"
-  ]
-}
\ No newline at end of file
+    "name": "stream-analytics",
```
Comment (ferhatelmas) [inline bower.json:2]: we could drop bower since we did for stream-js

File: src/specs.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    group: string;
+    value: string;
+};
+
+export type Engagement = {
+    content: string | Record<string, unknown>;
+    label: string;
+    boost?: number;
+    features?: Feature[];
+    feed_id?: string;
```
Comment (ferhatelmas) [inline src/specs.ts:11]: this is required

File: src/specs.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    score?: number;
+};
+
+export type Impression = {
+    content_list: Array<string | Record<string, unknown>>;
+    boost?: number;
+    features?: Feature[];
+    feed_id?: string;
+    location?: string;
+};
```
Comment (ferhatelmas) [inline src/specs.ts:23]: optional position is missing

Comment (ferhatelmas) [reply src/specs.ts:23]: otpional tracked_at (time) is missing

File: src/specs.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    features?: Feature[];
+    feed_id?: string;
+    location?: string;
+    position?: number;
+    score?: number;
+};
+
+export type Impression = {
+    content_list: Array<string | Record<string, unknown>>;
+    boost?: number;
```
Comment (ferhatelmas) [inline src/specs.ts:19]: boost isn't expected in impression

File: src/specs.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    features?: Feature[];
+    feed_id?: string;
```
Comment (ferhatelmas) [inline src/specs.ts:21]: this is required

File: src/specs.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    location?: string;
+    position?: number;
+    score?: number;
+};
+
+export type Impression = {
+    content_list: Array<string | Record<string, unknown>>;
```
Comment (ferhatelmas) [inline src/specs.ts:18]: not expected

File: src/stream-analytics.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    _sendEvent(resource: string, eventData: Impression | Engagement) {
+        if (this.userData === null) throw new errors.MissingUserId('userData was not set');
+
+        return request(`${this.baseUrl + resource}/?api_key=${this.apiKey}`, {
+            method: 'POST',
+            body: JSON.stringify({ ...eventData, user_data: this.userData }),
+            headers: {
+                'Content-Type': 'application/json',
+                'X-Stream-Client': `stream-javascript-client-${this.node ? 'node' : 'browser'}`,
```
Comment (ferhatelmas) [inline src/stream-analytics.ts:45]: can we add analytics and also version ?

File: src/specs.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    features?: Feature[];
+    feed_id?: string;
+    location?: string;
+};
+
+const validateFeatures = (features?: Feature[]) => {
+    if (!features) return '';
+    if (!Array.isArray(features)) return 'features should be array';
+
+    for (let i = 0; i < features.length; i += 1) {
```
Comment (ferhatelmas) [inline src/specs.ts:29]:
```js
for (let [group, value] of features) {
  if (!group || typeof group !== 'string') ...
  ....
}
```

might be better

File: src/specs.ts (hunk 2 above)
Comment (mahboubii) [reply src/specs.ts:11]: only `feed_id` is required?

File: src/specs.ts (hunk 8 above)
Comment (mahboubii) [reply src/specs.ts:29]: for..of falls into `no-restricted-syntax` of the linter

**Review by ferhatelmas — APPROVED** [review]
LGTM

**Pull request merged**